### PR TITLE
Tpetra: created Kokkos version for lowCommunicationMakeColMapAndReindex function

### PR DESCRIPTION
**Pull request by kxren1**
## Motivation
Developed to improve performance for transferAndFillComplete by using Kokkos in the function: lowCommunicationMakeColMapAndReindex
Addresses #11695

## Testing
Completed local performance testing, indicates Kokkos CUDA is more performant than previous host-based code

@jhux2 @rstumin @csiefer2 

WIP, to be completed:
- More testing on rectangular matrices, 
- Add logic to control usage of host-based code vs Kokkos code

### CODE REVIEW

**Comment by jhux2** [comment]
@trilinos/tpetra

File: packages/tpetra/core/src/Tpetra_Import_Util2.hpp [hunk 1]
```diff
@@ -946,6 +952,287 @@ lowCommunicationMakeColMapAndReindex (const Teuchos::ArrayView<const size_t> &ro
...
+    Kokkos::parallel_reduce(Kokkos::TeamThreadRange(member, rowptr_start, rowptr_end), [&](const size_t j, size_t &innerUpdate) {
+      const GO GID = colind_GID_view[j];
+      // Check if GID matches a row GID in local domain map
+      const LO LID = domainMap_local.getLocalElement (GID);
+      if (LID != -1) {
+        auto outcome = LocalGIDs_view_map.insert(LID);
+        // Fresh insert
+        if(outcome.success()) {
+          LocalGIDs_view[LID] = true;
+          Kokkos::atomic_add(&innerUpdate, 1);
```
Comment (brian-kelley) [inline packages/tpetra/core/src/Tpetra_Import_Util2.hpp:1033]: You don't have to use atomics to update partial reductions - Kokkos promises that no other thread gets the same reference at the same time

File: packages/tpetra/core/src/Tpetra_Import_Util2.hpp [hunk 2]
```diff
@@ -946,6 +952,287 @@ lowCommunicationMakeColMapAndReindex (const Teuchos::ArrayView<const size_t> &ro
...
+  
+  LO NumRemoteColGIDs = RemoteGIDs_view_map.size();
+
+  // For each index in RemoteGIDs_map that contains a GID, use valid_key_index[i] to indicate the number of GIDs "before" this GID
+  // Each valid_key_index[i], where i is an index in RemoteGIDs_map, is unique and consecutive
+  // This maps each element in the RemoteGIDs hash table to an index in RemoteGIDList / PIDList without any overwrites or empty spaces
+  Kokkos::View<int*> valid_key_index("valid_key_index", RemoteGIDs_view_map.capacity()); 
+  Kokkos::parallel_scan(RemoteGIDs_view_map.capacity(), KOKKOS_LAMBDA(const int i, GO& update, const bool final) {
+    if(final && RemoteGIDs_view_map.valid_at(i)) {
+      valid_key_index[i] = update;
```
Comment (brian-kelley) [inline packages/tpetra/core/src/Tpetra_Import_Util2.hpp:1056]: You can actually merge this with the parallel_for below (and not use ``valid_key_index``). ``final`` will be true exactly once for each index, so it's safe to do these here:
```
RemoteGIDList_view(update) = RemoteGIDs_view_map.key_at(i);
PIDList_view(update) = RemoteGIDs_view_map.value_at(i);
```

**Comment by kxren1** [comment]
Not ready for final review; ready for pre-test inspection

**Review by brian-kelley — DISMISSED** [review]

**Comment by brian-kelley** [comment]
@kxren1 Yes, all changes have to be tested and reviewed

**Review by brian-kelley — APPROVED** [review]

**Comment by brian-kelley** [comment]
@kxren1 This is ready to merge - I think you can click it, but if not I can do it

**Comment by kxren1** [comment]
@brian-kelley I'm not authorized to merge the PR myself - I'd greatly appreciate if you could do it. Thanks!

**Pull request merged**